### PR TITLE
Master mdl 72397 managecategories uienhancement 148 peerreviewfixes

### PR DESCRIPTION
**Pull request by marcghaly**
Peer review fixes from https://tracker.moodle.org/browse/MDL-72397 comments.

### CODE REVIEW

**Comment by safatshahin** [comment]
Hi @marcghaly 
I have seen there is a tracker review going on, can you please inform me what needs to be reviewed internally for this PR? unit tests are failing as well. Please close if the tracker review in progress.
Cheers!

**Comment by marcghaly** [comment]
Hi @safatshahin, there s currently a review that has been done, second comments were added, it is still having changes in progress, I will inform you as soon as changes are complete.

Thank you

File: question/bank/managecategories/classes/question_category_list_item.php [hunk 1]
```diff
@@ -90,15 +98,79 @@ public function item_html($extraargs = []) : string {
...
+                }
+            }
+        }
 
-        // Don't allow delete if this is the top category, or the last editable category in this context.
-        $deleteurl = null;
-        if ($category->parent && !helper::question_is_only_child_of_top_category_in_context($category->id)) {
-            $deleteurl = new moodle_url($this->parentlist->pageurl, ['delete' => $this->id, 'sesskey' => sesskey()]);
+        // Sets up export to XML link.
+        if (class_exists('\\qbank_exportquestions\\form\\export_form')) {
```
Comment (safatshahin) [inline question/bank/managecategories/classes/question_category_list_item.php:146]: need the plugin enable and disable check with this class check, class is for regression issues, plugin check is for plugin enable or disable

File: question/bank/managecategories/classes/question_category_object.php [hunk 2]
```diff
@@ -140,22 +154,34 @@ public function initialize($page, $contexts, $currentcat, $defaultcategory, $tod
         if (!$currentcat) {
             $this->catform->set_data(['parent' => $defaultcategory]);
         }
+        // Checkbox Form.
+        $checked = get_user_preferences('question_bank_qbshowdescr');
```
Comment (safatshahin) [inline question/bank/managecategories/classes/question_category_object.php:158]: just a thought, shouldnt it be a bit more specific? naming seems related to qbank base where is related to category description, can be named something better if its not too much work

File: question/bank/managecategories/classes/question_category_object.php [hunk 3]
```diff
@@ -213,6 +244,8 @@ public function get_course_ids(array $categories): array {
      * @param int $categoryid Category id.
      */
     public function edit_single_category(int $categoryid): void {
```
Comment (safatshahin) [inline question/bank/managecategories/classes/question_category_object.php:246]: mark for final deprecation and add the deprecation docs

File: question/bank/managecategories/classes/question_category_object.php [hunk 4]
```diff
@@ -416,6 +449,8 @@ public function add_category($newparent, $newcategory, $newinfo, $return = false
      */
     public function update_category($updateid, $newparent, $newname, $newinfo, $newinfoformat = FORMAT_HTML,
```
Comment (safatshahin) [inline question/bank/managecategories/classes/question_category_object.php:450]: mark for final deprecation and add the deprecation docs

File: question/bank/managecategories/tests/behat/category_reorder.feature [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+        | Course          | C1             | Course category   |
+        | Course          | C1             | Course category 2 |
+        | Category        | CAT1           | Category cat      |
+        | System          | S1             | System category   |
+    And I am on the "Test quiz Q001" "quiz activity" page logged in as "admin"
+    And I navigate to "Question bank > Categories" in current page administration
+
+  @javascript
+  Scenario: Teacher cannot move or delete single category under context
+    When I click on "//a[contains(text(), 'Default for Category 1')]/../../div[@class='float-right']/div[@class='float-left']" "xpath_element"
```
Comment (safatshahin) [inline question/bank/managecategories/tests/behat/category_reorder.feature:36]: my brain hurts looking at this xpath, please create a generator and possibly pass the Text (eg Default for Category 1) if its a standard one which looks like for the rest of the tests

File: question/bank/managecategories/tests/behat/category_reorder.feature [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+    Then I should not see "Delete"
+
+  @javascript
+  Scenario: Teacher can see complete edit menu if multiples categories exist under context
+    When I click on "//a[contains(text(), 'Quiz category')]/../../div[@class='float-right']/div[@class='float-left']" "xpath_element"
```
Comment (safatshahin) [inline question/bank/managecategories/tests/behat/category_reorder.feature:41]: generator here

File: question/bank/managecategories/tests/behat/category_reorder.feature [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+    Then I should see "Edit settings"
+    And I should see "Delete"
+    And I should see "Export as Moodle XML"
+
+  @javascript
+  Scenario: Teacher can reorder categories
+    When I drag "//a[contains(text(), 'Quiz category')]/../../div[@class='float-right']/div[@class='float-right']/span" "xpath_element" and I drop it in "Default for Category 1" "text"
```
Comment (safatshahin) [inline question/bank/managecategories/tests/behat/category_reorder.feature:48]: generator here and the rest like this

File: question/bank/managecategories/classes/external/update_question_category.php [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+                                            'info' => $info,
+                                            'infoformat' => $infoformat,
+                                            'idnumber' => $idnumber,
+                                            'id' => $id]);
+
+        $context = context_system::instance();
+        self::validate_context($context);
+        require_capability('moodle/category:manage', $context);
+
+        $updateid = $id;
```
[the remark this answers is not in the document]
Comment (marcghaly) [reply question/bank/managecategories/classes/external/update_question_category.php:85]: clean params added in the three web services.

File: question/bank/managecategories/classes/question_category_object.php (hunk 2 above)
Comment (marcghaly) [reply question/bank/managecategories/classes/question_category_object.php:158]: name changed to `qbank_managecategories_showdescr`

File: question/bank/managecategories/tests/behat/category_reorder.feature (hunk 6 above)
Comment (marcghaly) [reply question/bank/managecategories/tests/behat/category_reorder.feature:41]: Replaced with created step

File: question/bank/managecategories/tests/behat/category_reorder.feature (hunk 7 above)
Comment (marcghaly) [reply question/bank/managecategories/tests/behat/category_reorder.feature:48]: Replaced with created step

**Comment by safatshahin** [comment]
This tracker has gone for another peer review.
Closing.

**Pull request closed without merging**